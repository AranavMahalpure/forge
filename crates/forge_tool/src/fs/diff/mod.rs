mod apply;

pub use apply::{FSReplace, FSReplaceInput};