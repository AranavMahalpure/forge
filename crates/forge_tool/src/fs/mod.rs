mod file_info;
mod fs_find;
mod fs_list;
mod fs_read;
mod fs_replace;
mod fs_write;
mod syn;

pub use file_info::*;
pub use fs_find::*;
pub use fs_list::*;
pub use fs_read::*;
pub use fs_replace::*;
pub use fs_write::*;
