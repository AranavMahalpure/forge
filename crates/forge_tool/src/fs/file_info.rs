use forge_tool_macros::Description as DescriptionDerive;
use schemars::JsonSchema;
use serde::Deserialize;

use crate::{Description, ToolTrait};

#[derive(Deserialize, JsonSchema)]
pub struct FSFileInfoInput {
    pub path: String,
}

/// Retrieve detailed metadata about a file or directory. Returns comprehensive
/// information including size, creation time, last modified time, permissions,
/// and type. This tool is perfect for understanding file characteristics
/// without reading the actual content. Only works within allowed directories.
#[derive(DescriptionDerive)]
pub struct FSFileInfo;

#[async_trait::async_trait]
impl ToolTrait for FSFileInfo {
    type Input = FSFileInfoInput;
    type Output = String;

    async fn call(&self, input: Self::Input) -> Result<Self::Output, String> {
        let meta = tokio::fs::metadata(input.path)
            .await
            .map_err(|e| e.to_string())?;
        Ok(format!("{:?}", meta))
    }
}

#[cfg(test)]
mod test {
    use tempfile::TempDir;
    use tokio::fs;

    use super::*;

    #[tokio::test]
    async fn test_fs_file_info_on_file() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.txt");
        fs::write(&file_path, "test content").await.unwrap();

        let fs_info = FSFileInfo;
        let result = fs_info
            .call(FSFileInfoInput { path: file_path.to_string_lossy().to_string() })
            .await
            .unwrap();

        assert!(result.contains("FileType"));
        assert!(result.contains("permissions"));
        assert!(result.contains("modified"));
    }

    #[tokio::test]
    async fn test_fs_file_info_on_directory() {
        let temp_dir = TempDir::new().unwrap();
        let dir_path = temp_dir.path().join("test_dir");
        fs::create_dir(&dir_path).await.unwrap();

        let fs_info = FSFileInfo;
        let result = fs_info
            .call(FSFileInfoInput { path: dir_path.to_string_lossy().to_string() })
            .await
            .unwrap();

        assert!(result.contains("FileType"));
        assert!(result.contains("permissions"));
        assert!(result.contains("modified"));
    }

    #[tokio::test]
    async fn test_fs_file_info_nonexistent() {
        let temp_dir = TempDir::new().unwrap();
        let nonexistent_path = temp_dir.path().join("nonexistent");

        let fs_info = FSFileInfo;
        let result = fs_info
            .call(FSFileInfoInput { path: nonexistent_path.to_string_lossy().to_string() })
            .await;

        assert!(result.is_err());
    }
}
