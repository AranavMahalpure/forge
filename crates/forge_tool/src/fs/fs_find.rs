use std::collections::HashSet;
use std::path::Path;

use forge_domain::{Description, ToolCallService};
use forge_tool_macros::Description;
use forge_walker::Walker;
use schemars::JsonSchema;
use serde::Deserialize;

#[derive(Deserialize, JsonSchema)]
pub struct FSSearchInput {
    /// The path of the directory to search in (relative to the current working
    /// directory). This directory will be recursively searched.
    pub path: String,
    /// The regular expression pattern to search for. Uses Rust regex syntax.
    pub regex: String,
    /// Glob pattern to filter files (e.g., '*.ts' for TypeScript files). If not
    /// provided, it will search all files (*).
    pub file_pattern: Option<String>,
}

/// Request to perform a regex search across files in a specified directory,
/// providing context-rich results. This tool searches for patterns or specific
/// content across multiple files, displaying each match with encapsulating
/// context.
#[derive(Description)]
pub struct FSSearch;

#[async_trait::async_trait]
impl ToolCallService for FSSearch {
    type Input = FSSearchInput;
    type Output = Vec<String>;

    async fn call(&self, input: Self::Input) -> Result<Self::Output, String> {
        use regex::Regex;

        let dir = Path::new(&input.path);
        if !dir.exists() {
            return Err(format!("Directory '{}' does not exist", input.path));
        }

        // Create regex pattern - case-insensitive by default
        let pattern = format!("(?i){}", input.regex);
        let regex = Regex::new(&pattern).map_err(|e| format!("Invalid regex pattern: {}", e))?;

        let walker = Walker::new(dir.to_path_buf());
        let files = walker
            .get()
            .await
            .map_err(|e| format!("Failed to walk directory: {}", e))?;

        let mut matches = Vec::new();
        let mut seen_paths = HashSet::new();

        for file in files {
            if file.is_dir {
                continue;
            }

            let path = Path::new(&file.path);

            // Apply file pattern filter if provided
            if let Some(ref pattern) = input.file_pattern {
                let glob = glob::Pattern::new(pattern)
                    .map_err(|e| format!("Invalid glob pattern: {}", e))?;
                if let Some(filename) = path.file_name().and_then(|n| n.to_str()) {
                    if !glob.matches(filename) {
                        continue;
                    }
                }
            }

            // Skip if we've already processed this file
            if !seen_paths.insert(path.to_path_buf()) {
                continue;
            }

            // Full path for reading the file
            let full_path = dir.join(path);

            // Try to read the file content
            let content = match tokio::fs::read_to_string(&full_path).await {
                Ok(content) => content,
                Err(e) => {
                    // Skip binary or unreadable files silently
                    if e.kind() != std::io::ErrorKind::InvalidData {
                        matches.push(format!("Error reading {:?}: {}", path, e));
                    }
                    continue;
                }
            };

            // Process the file line by line
            for (line_num, line) in content.lines().enumerate() {
                if regex.is_match(line) {
                    // Format match in ripgrep style: filepath:line_num:content
                    matches.push(format!("{}:{}:{}", full_path.display(), line_num + 1, line));
                }
            }
        }

        if matches.is_empty() {
            Ok(vec![format!(
                "No matches found for pattern '{}' in path '{}'",
                input.regex, input.path
            )])
        } else {
            Ok(matches)
        }
    }
}

#[cfg(test)]
mod test {
    use tempfile::TempDir;
    use tokio::fs;

    use super::*;

    #[tokio::test]
    async fn test_fs_search_content() {
        let temp_dir = TempDir::new().unwrap();

        fs::write(temp_dir.path().join("test1.txt"), "Hello test world")
            .await
            .unwrap();
        fs::write(temp_dir.path().join("test2.txt"), "Another test case")
            .await
            .unwrap();
        fs::write(temp_dir.path().join("other.txt"), "No match here")
            .await
            .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "test".to_string(),
                file_pattern: None,
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 2);
        assert!(result.iter().any(|p| p.contains("test1.txt")));
        assert!(result.iter().any(|p| p.contains("test2.txt")));
    }

    #[tokio::test]
    async fn test_fs_search_with_pattern() {
        let temp_dir = TempDir::new().unwrap();

        fs::write(temp_dir.path().join("test1.txt"), "Hello test world")
            .await
            .unwrap();
        fs::write(temp_dir.path().join("test2.rs"), "fn test() {}")
            .await
            .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "test".to_string(),
                file_pattern: Some("*.rs".to_string()),
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 1);
        assert!(result.iter().any(|p| p.contains("test2.rs")));
    }

    #[tokio::test]
    async fn test_fs_search_with_context() {
        let temp_dir = TempDir::new().unwrap();
        let content = "line 1\nline 2\ntest line\nline 4\nline 5";

        fs::write(temp_dir.path().join("test.txt"), content)
            .await
            .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "test".to_string(),
                file_pattern: None,
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 1);
        assert!(result[0].contains("test line"));
    }

    #[tokio::test]
    async fn test_fs_search_recursive() {
        let temp_dir = TempDir::new().unwrap();

        let sub_dir = temp_dir.path().join("subdir");
        fs::create_dir(&sub_dir).await.unwrap();

        fs::write(temp_dir.path().join("test1.txt"), "test content")
            .await
            .unwrap();
        fs::write(sub_dir.join("test2.txt"), "more test content")
            .await
            .unwrap();
        fs::write(sub_dir.join("best.txt"), "this is proper\n test content")
            .await
            .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "test".to_string(),
                file_pattern: None,
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 3);
        assert!(result.iter().any(|p| p.contains("test1.txt")));
        assert!(result.iter().any(|p| p.contains("test2.txt")));
        assert!(result.iter().any(|p| p.contains("best.txt")));
    }

    #[tokio::test]
    async fn test_fs_search_case_insensitive() {
        let temp_dir = TempDir::new().unwrap();

        fs::write(
            temp_dir.path().join("test.txt"),
            "TEST CONTENT\ntest content",
        )
        .await
        .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "test".to_string(),
                file_pattern: None,
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 2);
        assert!(result.iter().any(|p| p.contains("TEST CONTENT")));
        assert!(result.iter().any(|p| p.contains("test content")));
    }

    #[tokio::test]
    async fn test_fs_search_no_matches() {
        let temp_dir = TempDir::new().unwrap();

        fs::write(temp_dir.path().join("test.txt"), "content")
            .await
            .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "nonexistent".to_string(),
                file_pattern: None,
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 1);
        assert!(result[0].contains("No matches found"));
    }

    #[tokio::test]
    async fn test_fs_search_invalid_regex() {
        let temp_dir = TempDir::new().unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "[invalid".to_string(),
                file_pattern: None,
            })
            .await;

        assert!(result.is_err());
        assert!(result.unwrap_err().contains("Invalid regex pattern"));
    }
}
