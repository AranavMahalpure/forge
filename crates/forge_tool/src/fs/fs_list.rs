use std::path::Path;

use anyhow::Context;
use forge_domain::{NamedTool, ToolCallService, ToolDescription, ToolName};
use forge_tool_macros::ToolDescription;
use forge_walker::Walker;
use schemars::JsonSchema;
use serde::Deserialize;

#[derive(Deserialize, JsonSchema)]
pub struct FSListInput {
    /// The path of the directory to list contents for (relative to the current
    /// working directory)
    pub path: String,
    /// Whether to list files recursively. Use true for recursive listing, false
    /// or omit for top-level only.
    pub recursive: Option<bool>,
}

/// Request to list files and directories within the specified directory. If
/// recursive is true, it will list all files and directories recursively. If
/// recursive is false or not provided, it will only list the top-level
/// contents. Do not use this tool to confirm the existence of files you may
/// have created, as the user will let you know if the files were created
/// successfully or not.
#[derive(ToolDescription)]
pub struct FSList;

impl NamedTool for FSList {
    fn tool_name(&self) -> ToolName {
        ToolName::new("tool_forge_fs_list")
    }
}

#[async_trait::async_trait]
impl ToolCallService for FSList {
    type Input = FSListInput;

    async fn call(&self, input: Self::Input) -> Result<String, String> {
        let dir = Path::new(&input.path);
        if !dir.exists() {
            return Err(format!("Directory '{}' does not exist", input.path));
        }

        let mut paths = Vec::new();
        let recursive = input.recursive.unwrap_or(false);
        let max_depth = if recursive { usize::MAX } else { 1 };
        let walker = Walker::new(dir.to_path_buf()).with_max_depth(max_depth);

        let files = walker
            .get()
            .await
            .with_context(|| format!("Failed to read directory contents from '{}'", input.path))
            .map_err(|e| e.to_string())?;

        for entry in files {
            // Skip the root directory itself
            if entry.path == dir.to_string_lossy() {
                continue;
            }

            if !entry.path.is_empty() {
                let prefix = if entry.is_dir { "[DIR]" } else { "[FILE]" };
                paths.push(format!("{} {}", prefix, entry.path));
            }
        }

        if paths.is_empty() {
            Ok("No files found".to_string())
        } else {
            Ok(paths.join("\n"))
        }
    }
}

#[cfg(test)]
mod test {
    use pretty_assertions::assert_eq;
    use tempfile::TempDir;
    use tokio::fs;

    use super::*;

    #[tokio::test]
    async fn test_fs_list_empty_directory() {
        let temp_dir = TempDir::new().unwrap();

        let fs_list = FSList;
        let result = fs_list
            .call(FSListInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                recursive: None,
            })
            .await
            .unwrap();

        assert_eq!(result, "No files found");
    }

    #[tokio::test]
    async fn test_fs_list_with_files_and_dirs() {
        let temp_dir = TempDir::new().unwrap();

        fs::write(temp_dir.path().join("file1.txt"), "content1")
            .await
            .unwrap();
        fs::write(temp_dir.path().join("file2.txt"), "content2")
            .await
            .unwrap();
        fs::create_dir(temp_dir.path().join("dir1")).await.unwrap();
        fs::create_dir(temp_dir.path().join("dir2")).await.unwrap();

        let fs_list = FSList;
        let result = fs_list
            .call(FSListInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                recursive: None,
            })
            .await
            .unwrap();

        let lines: Vec<_> = result.lines().collect();
        assert_eq!(lines.len(), 4);

        let files: Vec<_> = lines.iter().filter(|p| p.starts_with("[FILE]")).collect();
        let dirs: Vec<_> = lines.iter().filter(|p| p.starts_with("[DIR]")).collect();

        assert_eq!(files.len(), 2);
        assert_eq!(dirs.len(), 2);

        assert!(result.contains("file1.txt"));
        assert!(result.contains("file2.txt"));
        assert!(result.contains("dir1"));
        assert!(result.contains("dir2"));
    }

    #[tokio::test]
    async fn test_fs_list_nonexistent_directory() {
        let temp_dir = TempDir::new().unwrap();
        let nonexistent_dir = temp_dir.path().join("nonexistent");

        let fs_list = FSList;
        let result = fs_list
            .call(FSListInput {
                path: nonexistent_dir.to_string_lossy().to_string(),
                recursive: None,
            })
            .await;

        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_fs_list_with_hidden_files() {
        let temp_dir = TempDir::new().unwrap();

        fs::write(temp_dir.path().join("regular.txt"), "content")
            .await
            .unwrap();
        fs::write(temp_dir.path().join(".hidden"), "content")
            .await
            .unwrap();
        fs::create_dir(temp_dir.path().join(".hidden_dir"))
            .await
            .unwrap();

        let fs_list = FSList;
        let result = fs_list
            .call(FSListInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                recursive: None,
            })
            .await
            .unwrap();

        assert!(result.contains("regular.txt"));
        assert!(!result.contains(".hidden"));
        assert!(!result.contains(".hidden_dir"));
    }

    #[tokio::test]
    async fn test_fs_list_recursive() {
        let temp_dir = TempDir::new().unwrap();

        // Create nested directory structure
        fs::create_dir(temp_dir.path().join("dir1")).await.unwrap();
        fs::write(temp_dir.path().join("dir1/file1.txt"), "content1")
            .await
            .unwrap();
        fs::create_dir(temp_dir.path().join("dir1/subdir"))
            .await
            .unwrap();
        fs::write(temp_dir.path().join("dir1/subdir/file2.txt"), "content2")
            .await
            .unwrap();
        fs::write(temp_dir.path().join("root.txt"), "content3")
            .await
            .unwrap();

        let fs_list = FSList;

        // Test recursive listing
        let result = fs_list
            .call(FSListInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                recursive: Some(true),
            })
            .await
            .unwrap();

        let lines: Vec<_> = result.lines().collect();
        assert_eq!(lines.len(), 5); // root.txt, dir1, file1.txt, subdir, file2.txt
        assert!(result.contains("root.txt"));
        assert!(result.contains("dir1"));
        assert!(result.contains("file1.txt"));
        assert!(result.contains("subdir"));
        assert!(result.contains("file2.txt"));

        // Test non-recursive listing of same structure
        let result = fs_list
            .call(FSListInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                recursive: Some(false),
            })
            .await
            .unwrap();

        let lines: Vec<_> = result.lines().collect();
        assert_eq!(lines.len(), 2); // Only root.txt and dir1
        assert!(result.contains("root.txt"));
        assert!(result.contains("dir1"));
        assert!(!result.contains("file1.txt"));
        assert!(!result.contains("subdir"));
        assert!(!result.contains("file2.txt"));
    }
}
