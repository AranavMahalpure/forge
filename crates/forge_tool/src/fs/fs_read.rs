use forge_tool_macros::Description as DescriptionDerive;
use schemars::JsonSchema;
use serde::Deserialize;

use crate::{Description, ToolTrait};

#[derive(Deserialize, JsonSchema)]
pub struct FSReadInput {
    pub path: String,
}

/// Read the complete contents of a file from the file system. Handles various
/// text encodings and provides detailed error messages if the file cannot be
/// read. Use this tool when you need to examine the contents of a single file.
/// Only works within allowed directories.
#[derive(DescriptionDerive)]
pub struct FSRead;

#[async_trait::async_trait]
impl ToolTrait for FSRead {
    type Input = FSReadInput;
    type Output = String;

    async fn call(&self, input: Self::Input) -> Result<Self::Output, String> {
        let content = tokio::fs::read_to_string(&input.path)
            .await
            .map_err(|e| e.to_string())?;
        Ok(content)
    }
}

#[cfg(test)]
mod test {
    use tempfile::TempDir;
    use tokio::fs;

    use super::*;

    #[tokio::test]
    async fn test_fs_read_success() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.txt");

        let test_content = "Hello, World!";
        fs::write(&file_path, test_content).await.unwrap();

        let fs_read = FSRead;
        let result = fs_read
            .call(FSReadInput { path: file_path.to_string_lossy().to_string() })
            .await
            .unwrap();

        assert_eq!(result, test_content);
    }

    #[tokio::test]
    async fn test_fs_read_nonexistent_file() {
        let temp_dir = TempDir::new().unwrap();
        let nonexistent_file = temp_dir.path().join("nonexistent.txt");

        let fs_read = FSRead;
        let result = fs_read
            .call(FSReadInput { path: nonexistent_file.to_string_lossy().to_string() })
            .await;

        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_fs_read_empty_file() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("empty.txt");
        fs::write(&file_path, "").await.unwrap();

        let fs_read = FSRead;
        let result = fs_read
            .call(FSReadInput { path: file_path.to_string_lossy().to_string() })
            .await
            .unwrap();

        assert_eq!(result, "");
    }

    #[test]
    fn test_description() {
        assert!(FSRead::description().len() > 100)
    }
}
