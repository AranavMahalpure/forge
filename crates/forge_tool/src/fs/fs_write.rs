use forge_domain::{Description, ToolCallService};
use forge_tool_macros::Description;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::fs::syn;

#[derive(Deserialize, JsonSchema)]
pub struct FSWriteInput {
    /// The path of the file to write to (relative to the current working
    /// directory)
    pub path: String,
    /// The content to write to the file. ALWAYS provide the COMPLETE intended
    /// content of the file, without any truncation or omissions. You MUST
    /// include ALL parts of the file, even if they haven't been modified.
    pub content: String,
}

/// Request to write content to a file at the specified path. If the file
/// exists, it will be overwritten with the provided content. If the file
/// doesn't exist, it will be created. This tool will automatically create any
/// directories needed to write the file.
#[derive(Description)]
pub struct FSWrite;

#[async_trait::async_trait]
impl ToolCallService for FSWrite {
    type Input = FSWriteInput;
    type Output = FSWriteOutput;

    async fn call(&self, input: Self::Input) -> Result<Self::Output, String> {
        // Validate file content if it's a supported language file
        let syntax_checker = syn::validate(&input.path, &input.content).err();

        // Write file only after validation passes
        tokio::fs::write(&input.path, &input.content)
            .await
            .map_err(|e| e.to_string())?;

        Ok(FSWriteOutput { path: input.path, syntax_checker, content: input.content })
    }
}

#[derive(Serialize, JsonSchema)]
pub struct FSWriteOutput {
    pub path: String,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub syntax_checker: Option<String>,
}

#[cfg(test)]
mod test {
    use pretty_assertions::assert_eq;
    use tempfile::TempDir;
    use tokio::fs;

    use super::*;

    #[tokio::test]
    async fn test_fs_write_success() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.txt");

        let fs_write = FSWrite;
        let output = fs_write
            .call(FSWriteInput {
                path: file_path.to_string_lossy().to_string(),
                content: "Hello, World!".to_string(),
            })
            .await
            .unwrap();
        assert_eq!(output.path, file_path.to_string_lossy().to_string());
        assert_eq!(output.content, "Hello, World!");

        // Verify file was actually written
        let content = fs::read_to_string(&file_path).await.unwrap();
        assert_eq!(content, "Hello, World!")
    }

    #[tokio::test]
    async fn test_fs_write_invalid_rust() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.rs");

        let fs_write = FSWrite;
        let result = fs_write
            .call(FSWriteInput {
                path: file_path.to_string_lossy().to_string(),
                content: "fn main() { let x = ".to_string(),
            })
            .await;

        assert!(result.is_err());
        // File should not exist since validation failed
        assert!(!file_path.exists());
    }

    #[tokio::test]
    async fn test_fs_write_valid_rust() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.rs");

        let fs_write = FSWrite;
        let result = fs_write
            .call(FSWriteInput {
                path: file_path.to_string_lossy().to_string(),
                content: "fn main() { let x = 42; }".to_string(),
            })
            .await;

        assert!(result.is_ok());
        // Verify file contains valid Rust code
        let content = fs::read_to_string(&file_path).await.unwrap();
        assert_eq!(content, "fn main() { let x = 42; }");
    }
}
