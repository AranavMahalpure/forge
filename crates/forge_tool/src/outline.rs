use std::collections::{HashMap, HashSet};
use std::path::Path;

use forge_domain::{NamedTool, ToolCallService, ToolDescription, ToolName};
use forge_tool_macros::ToolDescription;
use schemars::JsonSchema;
use serde::Deserialize;
use streaming_iterator::{IntoStreamingIterator, StreamingIterator};
use tokio::fs;
use tree_sitter::{Language, Parser, Query, QueryCursor};
use walkdir::WalkDir;

const JAVASCRIPT: &str = include_str!("queries/javascript.rkt");
const PYTHON: &str = include_str!("queries/python.rkt");
const RUST: &str = include_str!("queries/rust.rkt");
const TYPESCRIPT: &str = include_str!("queries/typescript.rkt");
const CSS: &str = include_str!("queries/css.rkt");
const JAVA: &str = include_str!("queries/java.rkt");
const SCALA: &str = include_str!("queries/scala.rkt");

fn load_language_parser(language_name: &str) -> Result<Language, String> {
    match language_name {
        "rust" => Ok(tree_sitter_rust::LANGUAGE.into()),
        "javascript" => Ok(tree_sitter_javascript::LANGUAGE.into()),
        "python" => Ok(tree_sitter_python::LANGUAGE.into()),
        "typescript" => Ok(tree_sitter_typescript::LANGUAGE_TYPESCRIPT.into()),
        "tsx" => Ok(tree_sitter_typescript::LANGUAGE_TSX.into()),
        "css" => Ok(tree_sitter_css::LANGUAGE.into()),
        "java" => Ok(tree_sitter_java::LANGUAGE.into()),
        "scala" => Ok(tree_sitter_scala::LANGUAGE.into()),
        x => Err(format!("Unsupported language: {}", x)),
    }
}

fn load_queries() -> HashMap<&'static str, &'static str> {
    let mut queries = HashMap::new();
    queries.insert("rust", RUST);
    queries.insert("javascript", JAVASCRIPT);
    queries.insert("python", PYTHON);
    queries.insert("typescript", TYPESCRIPT);
    queries.insert("tsx", TYPESCRIPT); // Use TypeScript query for TSX files
    queries.insert("css", CSS);
    queries.insert("java", JAVA);
    queries.insert("scala", SCALA);
    queries
}

fn parse_file(_file: &Path, content: &str, parser: &mut Parser, query: &Query) -> Option<String> {
    let tree = parser.parse(content, None)?;
    let mut cursor = QueryCursor::new();
    let mut formatted_output = String::new();
    let mut last_line: i64 = -1;
    let mut seen_lines = HashSet::new();

    let mut captures: Vec<_> = cursor
        .matches(query, tree.root_node(), content.as_bytes())
        .flat_map(|m| m.captures.into_streaming_iter())
        .filter_map(|capture| {
            let node = capture.node;
            let start_line = node.start_position().row;
            // let end_line = node.end_position().row;
            // Get the full text of the node instead of just the first line
            let text = node.utf8_text(content.as_bytes()).ok()?;
            // Get the first line of the definition which contains the signature
            let first_line = text.lines().next()?.trim().to_string();
            Some((start_line, first_line))
        })
        .fold(Vec::default(), |mut acc, x| {
            acc.push(x.to_owned());
            acc
        });

    captures.sort_by_key(|&(row, _)| row);

    for (start_line, text) in captures {
        let start_line = start_line.to_owned() as i64;
        if !seen_lines.insert(start_line) {
            continue;
        }

        if last_line != -1 && start_line > last_line + 1 {
            formatted_output.push_str("|----\n");
        }

        formatted_output.push_str(&format!("â”‚{}\n", text.trim()));
        last_line = start_line;
    }

    if formatted_output.is_empty() {
        None
    } else {
        Some(formatted_output)
    }
}

#[derive(Deserialize, JsonSchema)]
pub struct OutlineInput {
    /// The path to the directory containing the source code files to analyze.
    pub path: String,
}

/// This tool helps developers analyze source code by listing key definitions
/// like classes, functions, and methods, making it easier to understand code
/// structure and relationships. It's ideal for navigating large or unfamiliar
/// codebases, improving comprehension during onboarding, code reviews, and
/// refactoring. The tool visualizes inheritance hierarchies, identifies
/// implementations, and uncovers dependencies and architectural patterns.
/// It tracks type usage, locates definitions, and clarifies module
/// organization, providing deep insights into system interactions. Supporting
/// most programming languages, it enhances productivity in complex projects and
/// notifies users when encountering unsupported languages.
#[derive(ToolDescription)]
pub struct Outline;

impl NamedTool for Outline {
    fn tool_name(&self) -> ToolName {
        ToolName::new("outline")
    }
}

#[async_trait::async_trait]
impl ToolCallService for Outline {
    type Input = OutlineInput;
    type Output = String;

    async fn call(&self, input: Self::Input) -> Result<Self::Output, String> {
        let extensions_to_languages = HashMap::from([
            ("rs", "rust"),
            ("js", "javascript"),
            ("py", "python"),
            ("ts", "typescript"),
            ("tsx", "tsx"),
            ("css", "css"),
            ("scss", "css"),
            ("java", "java"),
            ("scala", "scala"),
        ]);

        let queries = load_queries();
        let mut parsers: HashMap<&str, (Parser, Query)> = HashMap::new();
        let mut result = String::new();

        let entries = WalkDir::new(&input.path)
            .into_iter()
            .filter_map(|e| e.ok())
            .filter(|e| {
                e.file_type().is_file()
                    && e.path()
                        .extension()
                        .and_then(|e| e.to_str())
                        .map(|ext| {
                            extensions_to_languages.contains_key(ext.to_lowercase().as_str())
                        })
                        .unwrap_or(false)
            })
            .collect::<Vec<_>>();

        for entry in entries {
            let path = entry.path().to_path_buf();
            if let Ok(content) = fs::read_to_string(&path).await {
                if let Some(ext) = path.extension().and_then(|e| e.to_str()) {
                    if let Some(&lang_name) =
                        extensions_to_languages.get(ext.to_lowercase().as_str())
                    {
                        if !parsers.contains_key(lang_name) {
                            let language = load_language_parser(lang_name)?;
                            let mut parser = Parser::new();
                            parser.set_language(&language).map_err(|e| e.to_string())?;
                            let query = Query::new(&language, queries[lang_name])
                                .map_err(|e| e.to_string())?;
                            parsers.insert(lang_name, (parser, query));
                        }

                        if let Some((parser, query)) = parsers.get_mut(lang_name) {
                            if let Some(file_output) = parse_file(&path, &content, parser, query) {
                                if !result.is_empty() {
                                    result.push_str("|----\n");
                                }
                                result.push_str(&format!(
                                    "{}\n",
                                    path.file_name().unwrap().to_string_lossy()
                                ));
                                result.push_str(&file_output);
                            }
                        }
                    }
                }
            }
        }

        if result.is_empty() {
            Ok("No source code definitions found.".into())
        } else {
            Ok(result)
        }
    }
}

#[cfg(test)]
mod tests;
