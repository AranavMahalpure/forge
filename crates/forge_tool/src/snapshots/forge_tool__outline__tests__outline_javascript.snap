---
source: crates/forge_tool/src/outline.rs
expression: result
---
test.js
│function calculateTotal(items) {
|----
│class ShoppingCart {
|----
│addItem(item) {
